### PR TITLE
feat(infrastructure): Upload compiled screenshot test assets to GCS

### PR DESCRIPTION
**Pull request by acdvorak**
New NPM command: `screenshot:upload-assets`.

First, run:

```bash
npm start
```

Then, in another terminal, run:

```bash
export MDC_GCLOUD_SERVICE_ACCOUNT_KEY_FILE_PATH=path/to/gcloud-service-account-key.json
npm run screenshot:upload-assets
```

To avoid collisions between developers, every batch of files is uploaded to a directory with the following format:

```
/<bucket>/<username>/<timestamp>/<commit>/assets/
```

E.g.: `/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/`.

Sample output:

* [`mdc-button/classes/baseline.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/classes/baseline.html)
* [`mdc-button/classes/dense.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/classes/dense.html)
* [`mdc-button/mixins/container-fill-color.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/container-fill-color.html)
* [`mdc-button/mixins/corner-radius.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/corner-radius.html)
* [`mdc-button/mixins/filled-accessible.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/filled-accessible.html)
* [`mdc-button/mixins/horizontal-padding-baseline.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/horizontal-padding-baseline.html)
* [`mdc-button/mixins/horizontal-padding-dense.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/horizontal-padding-dense.html)
* [`mdc-button/mixins/icon-color.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/icon-color.html)
* [`mdc-button/mixins/ink-color.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/ink-color.html)
* [`mdc-button/mixins/stroke-color.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/stroke-color.html)
* [`mdc-button/mixins/stroke-width.html`](https://storage.googleapis.com/mdc-web-screenshot-tests/advorak/2018/04/03/23_00_32_993/5ca514fe/assets/mdc-button/mixins/stroke-width.html)

### CODE REVIEW

File: test/screenshot/asset-uploader.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    error: null,
+    bucketUrl: GCLOUD_STORAGE_BASE_URL,
+    parentDir: absoluteGcsBaseDir,
+    relativePath: relativeGcsFilePath,
+    fullUrl: `${GCLOUD_STORAGE_BASE_URL}${absoluteGcsBaseDir}${relativeGcsFilePath}`,
+  };
+}
+
+function handleUploadFailure(absoluteGcsBaseDir, relativeGcsFilePath, err) {
+  console.error(`❌︎ FAILED to upload ${GCLOUD_STORAGE_BASE_URL}${absoluteGcsBaseDir}${relativeGcsFilePath}:\n`, err);
```
Comment (patrickrodee) [inline test/screenshot/asset-uploader.js]: What icon is this? I only see a square.

File: test/screenshot/run-upload-assets.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    (files) => {
+      const htmlFileUrls =
+        files
+          .filter((file) => file.relativePath.endsWith('.html'))
+          .map((file) => file.fullUrl)
+          .sort();
+      console.log('\n\nDONE!\n\n');
+      console.log(htmlFileUrls.join('\n'));
+    },
+    (err) => {
```
Comment (patrickrodee) [inline test/screenshot/run-upload-assets.js]: Nit: Use `.catch` instead. It's 100% a personal preference but I find it much more obvious what the function is doing.

File: test/screenshot/asset-uploader.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  // Attaching Git metadata makes it easier to track down the source of uploaded files.
+  const gitLocalBranchName = await git('revparse', ['--abbrev-ref', 'HEAD']);
+  const gitBaseCommitFull = await git('revparse', ['HEAD']);
+  const gitBaseCommitShort = shorten(gitBaseCommitFull);
+
+  const absoluteGcsBaseDir = `${USERNAME}/${gitBaseCommitShort}/${uuidShort}/assets/`;
+  const relativeLocalFilePaths = glob.sync(path.join(LOCAL_DIRECTORY_PREFIX, '**/*.*'));
+
+  relativeLocalFilePaths.forEach((relativeLocalFilePath) => {
+    const fileContents = fs.readFileSync(relativeLocalFilePath);
```
[the remark this answers is not in the document]
Comment (acdvorak) [reply test/screenshot/asset-uploader.js]: No, it's to ensure that `upload()` doesn't immediately return an empty array.

I tried several different approaches first, but they were all more convoluted and harder to read, so I went with `readFileSync()` cuz it was the simplest.

File: test/screenshot/run-upload-assets.js (hunk 2 above)
Comment (acdvorak) [reply test/screenshot/run-upload-assets.js]: Done

File: test/screenshot/asset-uploader.js (hunk 1 above)
Comment (acdvorak) [reply test/screenshot/asset-uploader.js]: Good catch! Changed to ✗ ([Ballot X](http://www.fileformat.info/info/unicode/char/2717/index.htm))

File: test/screenshot/asset-uploader.js [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+          'X-MDC-Git-Commit': gitCommitShort,
+          'X-MDC-Git-Branch': gitBranchName,
+        },
+      },
+    };
+
+    promises.push(
+      file.save(fileContents, fileOptions).then(
+        () => handleUploadSuccess(baseGcsDir, relativeGcsFilePath),
+        (err) => handleUploadFailure(baseGcsDir, relativeGcsFilePath, err)
```
Comment (patrickrodee) [inline test/screenshot/asset-uploader.js]: SUPER NIT: use `.catch` here, too. Gonna LGTM this without because this is just me being obscenely picky.

Comment (acdvorak) [reply test/screenshot/asset-uploader.js]: Good catch! Done.

**Pull request merged**